spark/src/main/scala/geotrellis/spark/clip/ClipToGrid.scala

```diff
@@ -16,191 +16,111 @@
-        clipGeom(e, f.geom, p).map(Feature(_, ()))
-      }
-    ).mapValues(_.geom)
+    rdd: RDD[G],
+    clipGeom: (Extent, Feature[G, Unit]) => Option[Feature[Geometry, Unit]]
+  ): RDD[(SpatialKey, Geometry)] =
+    featuresWith(layout, rdd.map(Feature(_, ())), clipGeom).mapValues(_.geom)
 
   /** Clip each geometry in the RDD to the set of SpatialKeys
     * which intersect it, where the SpatialKeys map to the
     * given [[LayoutDefinition]], using the given method
     * to clip each geometry to the extent.
     */
-  def apply[G <: Geometry, D](
-    rdd: RDD[Feature[G, D]],
-    layout: LayoutDefinition
+  def features[G <: Geometry, D](
+    layout: LayoutDefinition,
+    rdd: RDD[Feature[G, D]]
   ): RDD[(SpatialKey, Feature[Geometry, D])] =
-    apply[G, D](rdd, layout, clipFeatureToExtent[G,D] _)
+    featuresWith(layout, rdd, byExtent)
 
   /** Clip each geometry in the RDD to the set of SpatialKeys
     * which intersect it, where the SpatialKeys map to the
     * given [[LayoutDefinition]], using the given method
     * to clip each geometry to the extent.
     */
-  def apply[G <: Geometry, D](
-    rdd: RDD[Feature[G, D]],
+  def featuresWith[G <: Geometry, D](
     layout: LayoutDefinition,
-    clipFeature: (Extent, Feature[G, D], Predicates) => Option[Feature[Geometry, D]]
+    rdd: RDD[Feature[G, D]],
+    clipFeature: (Extent, Feature[G, D]) => Option[Feature[Geometry, D]]
   ): RDD[(SpatialKey, Feature[Geometry, D])] = {
     val mapTransform: MapKeyTransform = layout.mapTransform
 
-    rdd.flatMap { f => clipGeom(mapTransform, clipFeature, f) }
+    /* Associate each Feature with its SpatialKeys*/
+    val withKeys: RDD[(Iterator[SpatialKey], Feature[G,D])] =
+      rdd.map(f => (mapTransform.keysForGeometry(f.geom), f))
+
+    /* Clip every Feature */
+    withKeys.flatMap { case (ks, f) => ks.flatMap(k => clipFeature(k.extent(layout), f).map((k, _))) }
+  }
+
+  /** Clips Features to fit the given Extent. Does its best to avoid unnecessary
+    * JTS interactions.
+    */
+  def byExtent[G <: Geometry, D](extent: Extent, f: Feature[G, D]): Option[Feature[Geometry, D]] = f.geom match {
+    case g: Point                              => Some(f)
+    case g: MultiPoint if coveredBy(g, extent) => Some(f)
+    case g: MultiPoint                         => clip(extent, g).map(Feature(_, f.data))
+    case g: Line       if coveredBy(g, extent) => Some(f)
+    case g: Line                               => clip(extent, g).map(Feature(_, f.data))
+    case g: MultiLine  if coveredBy(g, extent) => Some(f)
+    case g: MultiLine                          => clip(extent, g).map(Feature(_, f.data))
+    case g =>  /* Polygon and MultiPolygon */
+      val pg = PreparedGeometryFactory.prepare(g.jtsGeom)
+      val ep = extent.toPolygon.jtsGeom
+
+      if (pg.covers(ep)) Some(Feature(extent, f.data))
+      else if (pg.coveredBy(ep)) Some(f)
+      else clip(extent, g).map(Feature(_, f.data))
```

Comment:
The logic contained in `Predicates` was considered the best reasonable approach for the given Geoms, so the same checks are now hard-coded into the canonical clipping function `byExtent`.

Reply:
This is a de-optimization.

Building a prepared Geometry takes time. That is why the original code was only doing it once for each of the clips it was going to do on that Geometry.

The predicates also apply across the board no matter what clipping method you use, so I don't think they should be hidden in a single implementation. That's why I elevated them out of the clipping logic and passed into the clip function, so that any clip function can take advantage of the predicates.

Reply:
Good catch. The first thing that comes to mind is switching the signature of clipping functions to:

```scala
(Iterator[Extent], Feature[G, D]) => Iterator[Option[Feature[G, D]]]
```

since the new `MapKeyTransform.keysForGeometry` also returns an `Iterator`. The actual clipping (done in a `Iterator.map`) and the subsequent `foldLeft` to grab out the legal Features would be optimally fast due to the chained, lazy `Iterator` ops. VectorPipe would do the error checking in the `foldLeft`, and vanilla GT can throw as it sees fit.

I had a question for you for when you got back to the office: was the intersection meant to be done on the `PreparedGeometry`, or just the `covers` check? At current both the original `ClipToGrid` and this rewrite only use the `PreparedGeometry` for the `covers` check. Was that an oversight or am I misunderstanding?

Reply:
There was not an oversight. PreparedGeometry is only an optimization for the predicate, and not the intersection operation.

What's the reason for changing it from what it was? I'm unclear why this part of the code needed to change.

Reply:
Ok cool. As for the change, I also mentioned the reasons a bit [here](https://github.com/locationtech/geotrellis/pull/2453#discussion_r147570288). Integrating `ClipToGrid` was my immediate task for VP, but I hit issues. I read through your code a lot and  saw an isomorphic algorithm that didn't involve the `Predicates` concept, so I [prototyped it over on the VP side](https://github.com/geotrellis/vectorpipe/pull/29). I explained everything to the team and was given the go-ahead to port it back.